_posts/2018-09-23-Inter-service-communication.md

```diff
@@ -0,0 +1,208 @@
+It's important to to let the builder know which client he has to target for communication. 
+The second parameter is most likely the base url where all the requests begin. 
+
+{% highlight java %}
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+### Interceptor
+If you need some basic authorization, custom headers or some extra information in every request of the client, we can use interceptors. 
+This becomes very useful in situations where every requests needs this extra information.
+To add an interceptor, we just add an extra method that returns the Feign interceptor. 
+
+{% highlight java %}
+ private RequestInterceptor requestInterceptor() {
+        return requestTemplate -> {
+            requestTemplate.header("user", username);
+            requestTemplate.header("password", password);
+            requestTemplate.header("Accept", ContentType.APPLICATION_JSON.getMimeType());
+        };
+    }
+{% endhighlight %}
+
+To enable the customization, we add the interceptor to the builder. 
+
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .requestInterceptor(requestInterceptor())
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+### Client
+By default, the default HTTP client of Feign uses HttpUrlConnection to execute its HTTP requests.
+You can configure another client (ApacheHttpClient, OkHttpClient, ...) as follows:
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .client(new ApacheHttpClient())
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+#### OkHttpClient
+OkHttp is an HTTP client that’s efficient by default:
+
+* HTTP/2 support allows all requests to the same host to share a socket.
+* Connection pooling reduces request latency (if HTTP/2 isn’t available).
+* Transparent GZIP shrinks download sizes.
+* Response caching avoids the network completely for repeat requests.
+
+#### Mutual SSL with ApacheHttpClient
+The advantage of using ApacheHttpClient over the default client is that ApacheHttpClient sends more headers with the request, eg. Content-Length, which some servers expect.
+Aside, Apache brings support for mutual SSL, where we can safely store our key and trust-store in to transit.
+To achieve this, you have to create a custom ApacheHttpClient and insert it in to our Feign client. 
+
+
+### Retry mechanism
+When we want to build some resilience in our communication, we can setup a retry mechanism in our Feign client. 
+If the other service is unreachable, we will try again until it is healthy or until our configuration has been set. 
+When we want to use the retryer of Feign, we got 3 properties we can set.
+
+> Period: How long it takes before the retry is triggered
```

Comment:
Instead of the citation styling I would stick to the bullet list similar to when you explained the different properties of configuring Feign.